pkg/reconcilermanager/controllers/reconciler_base.go

```diff
@@ -617,12 +617,13 @@ func (r *reconcilerBase) setupOrTeardown(ctx context.Context, syncObj client.Obj
 	// Else - the object is being deleted.
 
 	if controllerutil.ContainsFinalizer(syncObj, metadata.ReconcilerFinalizer) {
-		// The object is being deleted, but the reconciler finalizer is still running.
-		// Wait for the reconciler finalizer to complete.
-		r.logger(ctx).Info("Waiting for Reconciler Finalizer to finish")
-		return nil
+		removed, err := r.removeReconcilerFinalizer(ctx, syncObj)
```

Comment:
This method name makes this confusing to read. It doesn't always remove the finalizer. Sometimes it just returns and waits for a new event to say it was removed.

Also, the function itself has different logic for the rootsync and reposync. Can we instead push the type-specific logic into the reposync and rootsync controllers? Maybe inject a finalizer strategy that implements an interface method?

It would be nice to keep the code in the base class common to both impls.

Reply:
sounds good. I'll change to use interface and type-specific implementation.

Reply:
Added a `reconcilerFinalizerHandler` interface and a `handleReconcilerFinalizer` function.
The name is a bit long and I'm open for better names.